File: src/kube_chainsaw/loader.py
"""YAML manifest loader for Kubernetes resources."""

import re
import sys
from pathlib import Path
from typing import List, Optional, Set, Union

import yaml

from .models import LoadedResources

# Default directories to exclude from scanning
DEFAULT_EXCLUDE_DIRS = {".git", "vendor", "node_modules", "bin"}

# Default limits
DEFAULT_MAX_FILE_SIZE = 10 * 1024 * 1024  # 10MB
DEFAULT_MAX_DOCS_PER_FILE = 10000


def _is_go_template(filepath: Path, content: str) -> bool:
    """Detect if file contains Go template syntax.

    Args:
        filepath: Path to the file.
        content: File content string.

    Returns:
        True if file is a Go template.
    """
    # Check filename
    filename = filepath.name
    if ".tmpl.yaml" in filename or ".template.yaml" in filename:
        return True

    # Check content for {{ }} patterns
    if "{{" in content:
        return True

    return False


def _preprocess_go_template(content: str) -> str:
    """Replace Go template expressions with placeholder values.

    Args:
        content: YAML content with Go templates.

    Returns:
        Content with templates replaced.
    """
    # Replace {{ ... }} with "placeholder-value"
    return re.sub(r"\{\{[^}]+\}\}", "placeholder-value", content)


def _should_exclude_path(filepath: Path, exclude_dirs: Set[str]) -> bool:
    """Check if file path should be excluded.

    Args:
        filepath: Path to check.
        exclude_dirs: Set of directory names to exclude.

    Returns:
        True if path should be excluded.
    """
    # Check if any part of the path matches exclude dirs
    for part in filepath.parts:
        if part in exclude_dirs:
            return True
    return False


def _process_document(
    doc: dict, filepath: Path, resources: LoadedResources
) -> None:
    """Process a single YAML document and categorize it.

    Args:
        doc: Parsed YAML document.
        filepath: Path to the source file.
        resources: LoadedResources to populate.
    """
    if not isinstance(doc, dict):
        return

    kind = doc.get("kind")
    metadata = doc.get("metadata", {})
    name = metadata.get("name")
    namespace = metadata.get("namespace", "default")

    if not kind or not name:
        return

    file_str = str(filepath)

    if kind == "ClusterRole":
        rules = doc.get("rules", [])
        resources.cluster_roles[name] = {
            "rules": rules,
            "file": file_str,
            "doc": doc,
        }

    elif kind == "Role":
        rules = doc.get("rules", [])
        key = f"{namespace}/{name}"
        resources.roles[key] = {
            "rules": rules,
            "file": file_str,
            "doc": doc,
        }

    elif kind == "ClusterRoleBinding":
        resources.cluster_role_bindings.append({
            "doc": doc,
            "file": file_str,
        })

    elif kind == "RoleBinding":
        resources.role_bindings.append({
            "doc": doc,
            "file": file_str,
        })

    elif kind == "ServiceAccount":
        key = f"{namespace}/{name}"
        resources.service_accounts[key] = {
            "file": file_str,
            "doc": doc,
        }

    elif kind == "Pod":
        spec = doc.get("spec", {})
        service_account = spec.get("serviceAccountName", spec.get("serviceAccount", "default"))
        automount_token = spec.get("automountServiceAccountToken", True)

        key = f"{namespace}/{name}"
        resources.pods[key] = {
            "serviceAccount": service_account,
            "file": file_str,
            "automountToken": automount_token,
            "doc": doc,
        }


def _process_file(
    filepath: Path,
    resources: LoadedResources,
    max_file_size: int,
    max_docs_per_file: int,
) -> None:
    """Process a single YAML file.

    Args:
        filepath: Path to YAML file.
        resources: LoadedResources to populate.
        max_file_size: Maximum file size in bytes.
        max_docs_per_file: Maximum documents per file.
    """
    # Check file size
    try:
        file_size = filepath.stat().st_size
        if file_size > max_file_size:
            print(
                f"WARNING: Skipping {filepath} (size {file_size} exceeds limit {max_file_size})",
                file=sys.stderr,
            )
            return
    except OSError as e:
        print(f"WARNING: Could not stat {filepath}: {e}", file=sys.stderr)
        return

    # Read file content
    try:
        content = filepath.read_text(encoding="utf-8")
    except (OSError, UnicodeDecodeError) as e:
        print(f"WARNING: Could not read {filepath}: {e}", file=sys.stderr)
        return

    # Check if Go template and preprocess
    if _is_go_template(filepath, content):
        content = _preprocess_go_template(content)

    # Parse YAML
    try:
        docs = list(yaml.safe_load_all(content))
    except yaml.YAMLError as e:
        # Only warn for non-template files
        if not _is_go_template(filepath, content):
            print(f"WARNING: YAML parsing error in {filepath}: {e}", file=sys.stderr)
        return

    # Process documents
    doc_count = 0
    for doc in docs:
        if doc is None:
            continue

        doc_count += 1
        if doc_count > max_docs_per_file:
            print(
                f"WARNING: Stopping processing of {filepath} after {max_docs_per_file} documents",
                file=sys.stderr,
            )
            break

        _process_document(doc, filepath, resources)


def load_manifests(
    paths: List[Union[str, Path]],
    exclude_dirs: Optional[Set[str]] = None,
    use_default_excludes: bool = True,
    max_file_size: int = DEFAULT_MAX_FILE_SIZE,
    max_docs_per_file: int = DEFAULT_MAX_DOCS_PER_FILE,
) -> LoadedResources:
    """Load Kubernetes manifests from YAML files.

    Args:
        paths: List of file or directory paths to scan.
        exclude_dirs: Additional directories to exclude from scanning.
        use_default_excludes: Whether to use default exclude directories.
        max_file_size: Maximum file size in bytes (default 10MB).
        max_docs_per_file: Maximum documents per file (default 10000).

    Returns:
        LoadedResources containing categorized Kubernetes resources.
    """
    resources = LoadedResources()

    # Build exclude set
    exclude_set: Set[str] = set()
    if use_default_excludes:
        exclude_set.update(DEFAULT_EXCLUDE_DIRS)
    if exclude_dirs:
        exclude_set.update(exclude_dirs)

    # Process each path
    for path in paths:
        path_obj = Path(path)

        if not path_obj.exists():
            print(f"WARNING: Path does not exist: {path}", file=sys.stderr)
            continue

        if path_obj.is_file():
            # Process single file
            if not path_obj.is_symlink():
                _process_file(path_obj, resources, max_file_size, max_docs_per_file)
        else:
            # Scan directory
            for pattern in ["*.yaml", "*.yml"]:
                for filepath in path_obj.rglob(pattern):
                    # Skip symlinks
                    if filepath.is_symlink():
                        continue

                    # Skip excluded directories
                    if _should_exclude_path(filepath, exclude_set):
                        continue

                    # Skip if not a file (shouldn't happen with rglob, but be safe)
                    if not filepath.is_file():
                        continue

                    _process_file(filepath, resources, max_file_size, max_docs_per_file)

    return resources

File: pkg/loader/loader.go
package loader

import (
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/ugiordan/kube-chainsaw/pkg/models"
	"sigs.k8s.io/yaml"
)

const (
	DefaultMaxFileSize    = 10 * 1024 * 1024 // 10MB
	DefaultMaxDocsPerFile = 10000
)

// DefaultExcludeDirs contains directory names skipped during traversal.
var DefaultExcludeDirs = map[string]bool{
	".git":         true,
	"vendor":       true,
	"node_modules": true,
	"bin":          true,
}

// goTemplateRe matches Go template expressions like {{ .Values.foo }}.
var goTemplateRe = regexp.MustCompile(`\{\{[^}]+\}\}`)

// Options configures the manifest loader.
type Options struct {
	ExcludeDirs        []string
	UseDefaultExcludes bool
	MaxFileSize        int64
	MaxDocsPerFile     int
}

// DefaultOptions returns options with sensible defaults.
func DefaultOptions() *Options {
	return &Options{
		UseDefaultExcludes: true,
		MaxFileSize:        DefaultMaxFileSize,
		MaxDocsPerFile:     DefaultMaxDocsPerFile,
	}
}

// LoadManifests walks the given paths (files or directories) and parses all
// YAML/JSON manifests into LoadedResources.
func LoadManifests(paths []string, opts *Options) (*models.LoadedResources, error) {
	if opts == nil {
		opts = DefaultOptions()
	}
	if opts.MaxFileSize <= 0 {
		opts.MaxFileSize = DefaultMaxFileSize
	}
	if opts.MaxDocsPerFile <= 0 {
		opts.MaxDocsPerFile = DefaultMaxDocsPerFile
	}

	excludes := buildExcludes(opts)
	result := models.NewLoadedResources()

	for _, p := range paths {
		info, err := os.Lstat(p)
		if err != nil {
			return nil, fmt.Errorf("cannot access %q: %w", p, err)
		}

		// Skip symlinks at the top level
		if info.Mode()&fs.ModeSymlink != 0 {
			continue
		}

		if info.IsDir() {
			if err := walkDir(p, excludes, opts, result); err != nil {
				return nil, err
			}
		} else {
			if err := processFile(p, opts, result); err != nil {
				// Skip files that fail to parse (non-fatal)
				continue
			}
		}
	}

	return result, nil
}

func buildExcludes(opts *Options) map[string]bool {
	excludes := make(map[string]bool)
	if opts.UseDefaultExcludes {
		for k, v := range DefaultExcludeDirs {
			excludes[k] = v
		}
	}
	for _, d := range opts.ExcludeDirs {
		excludes[d] = true
	}
	return excludes
}

func walkDir(root string, excludes map[string]bool, opts *Options, result *models.LoadedResources) error {
	return filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil // skip inaccessible entries
		}

		// Skip symlinks
		if d.Type()&fs.ModeSymlink != 0 {
			if d.IsDir() {
				return filepath.SkipDir
			}
			return nil
		}

		if d.IsDir() {
			if excludes[d.Name()] && path != root {
				return filepath.SkipDir
			}
			return nil
		}

		if !isYAMLFile(path) {
			return nil
		}

		// processFile errors are non-fatal: skip malformed files
		_ = processFile(path, opts, result)
		return nil
	})
}

func isYAMLFile(path string) bool {
	ext := strings.ToLower(filepath.Ext(path))
	return ext == ".yaml" || ext == ".yml" || ext == ".json"
}

func processFile(path string, opts *Options, result *models.LoadedResources) error {
	info, err := os.Stat(path)
	if err != nil {
		return err
	}

	if info.Size() > opts.MaxFileSize {
		return fmt.Errorf("file %q exceeds max size (%d > %d)", path, info.Size(), opts.MaxFileSize)
	}

	data, err := os.ReadFile(path)
	if err != nil {
		return err
	}

	content := string(data)

	// Preprocess Go templates so the YAML parser doesn't choke
	content = goTemplateRe.ReplaceAllString(content, "placeholder-value")

	docs := splitYAMLDocs(content)

	if len(docs) > opts.MaxDocsPerFile {
		return fmt.Errorf("file %q has too many documents (%d > %d)", path, len(docs), opts.MaxDocsPerFile)
	}

	for _, doc := range docs {
		doc = strings.TrimSpace(doc)
		if doc == "" {
			continue
		}

		var parsed map[string]interface{}
		if err := yaml.Unmarshal([]byte(doc), &parsed); err != nil {
			continue // skip unparseable documents
		}

		if parsed == nil {
			continue
		}

		categorize(parsed, path, result)
	}

	return nil
}

// splitYAMLDocs splits a multi-document YAML string on "---" separators.
// Only splits on "---" that appears at the start of a line (not inside
// multiline strings, which the simple approach can't fully handle, but
// works well enough for Kubernetes manifests).
func splitYAMLDocs(content string) []string {
	// Split on document separator at the start of a line
	re := regexp.MustCompile(`(?m)^---\s*$`)
	return re.Split(content, -1)
}

func categorize(doc map[string]interface{}, file string, result *models.LoadedResources) {
	kind, _ := doc["kind"].(string)
	metadata, _ := doc["metadata"].(map[string]interface{})
	if metadata == nil {
		return
	}
	name, _ := metadata["name"].(string)
	namespace, _ := metadata["namespace"].(string)

	switch kind {
	case "ClusterRole":
		rules := extractRules(doc)
		result.ClusterRoles[name] = &models.ClusterRoleData{
			Rules: rules,
			File:  file,
			Doc:   doc,
		}

	case "Role":
		rules := extractRules(doc)
		key := namespace + "/" + name
		result.Roles[key] = &models.RoleData{
			Rules:     rules,
			Namespace: namespace,
			File:      file,
			Doc:       doc,
		}

	case "ClusterRoleBinding":
		bd := extractBinding(doc, name, "", file)
		result.ClusterRoleBindings = append(result.ClusterRoleBindings, bd)

	case "RoleBinding":
		bd := extractBinding(doc, name, namespace, file)
		result.RoleBindings = append(result.RoleBindings, bd)

	case "ServiceAccount":
		key := namespace + "/" + name
		result.ServiceAccounts[key] = &models.SAData{
			Name:      name,
			Namespace: namespace,
			File:      file,
			Doc:       doc,
		}

	case "Pod":
		saName := extractServiceAccountName(doc)
		key := namespace + "/" + name
		result.Pods[key] = &models.PodData{
			Name:               name,
			Namespace:          namespace,
			ServiceAccountName: saName,
			File:               file,
			Doc:                doc,
		}
	}
}

func extractRules(doc map[string]interface{}) []map[string]interface{} {
	rulesRaw, ok := doc["rules"]
	if !ok {
		return nil
	}

	rulesSlice, ok := rulesRaw.([]interface{})
	if !ok {
		return nil
	}

	var rules []map[string]interface{}
	for _, r := range rulesSlice {
		rMap, ok := r.(map[string]interface{})
		if ok {
			rules = append(rules, rMap)
		}
	}
	return rules
}

func extractBinding(doc map[string]interface{}, name, namespace, file string) *models.BindingData {
	bd := &models.BindingData{
		Name:      name,
		Namespace: namespace,
		File:      file,
		Doc:       doc,
	}

	if roleRef, ok := doc["roleRef"].(map[string]interface{}); ok {
		bd.RoleRef = roleRef
	}

	if subjects, ok := doc["subjects"].([]interface{}); ok {
		for _, s := range subjects {
			if sMap, ok := s.(map[string]interface{}); ok {
				bd.Subjects = append(bd.Subjects, sMap)
			}
		}
	}

	return bd
}

func extractServiceAccountName(doc map[string]interface{}) string {
	spec, ok := doc["spec"].(map[string]interface{})
	if !ok {
		return "default"
	}

	saName, ok := spec["serviceAccountName"].(string)
	if !ok || saName == "" {
		return "default"
	}
	return saName
}
